group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -544,6 +573,81 @@ public OffsetFetchResponseData.OffsetFetchResponseGroup fetchAllOffsets(
             .setTopics(topicResponses);
     }
 
+    /**
+     * Remove expired offsets for group.
+     *
+     * @param groupId The group id.
+     * @param records The list of records to populate with offset commit tombstone records.
+     * @param offsetsRetentionMs The offset retention in milliseconds.
+     *
+     * @return The group id if the group no longer has any offsets remaining, empty otherwise.
+     */
+    public Optional<String> cleanupExpiredOffsets(String groupId, List<Record> records, long offsetsRetentionMs) {
+        TimelineHashMap<String, TimelineHashMap<Integer, OffsetAndMetadata>> offsetsByTopic = offsetsByGroup.get(groupId);
+        if (offsetsByTopic == null) {
+            return Optional.of(groupId);
+        }
+        try {
+            Group group = groupMetadataManager.group(groupId);
+            ExpirationCondition expirationCondition = group.expirationCondition();
+            Set<TopicPartition> expiredPartitions = new HashSet<>();
+            long currentTimestamp = time.milliseconds();
+            AtomicBoolean hasAllOffsetsExpired = new AtomicBoolean(true);
+            offsetsByTopic.forEach((topic, partitions) -> {
+                if (!expirationCondition.subscribedTopics.contains(topic)) {
+                    partitions.forEach((partition, offsetAndMetadata) -> {
+                        OptionalLong expireTimestampMs = offsetAndMetadata.expireTimestampMs;
+                        if (expireTimestampMs.isPresent()) {
+                            // Older versions with explicit expire_timestamp field => old expiration semantics is used
+                            if (currentTimestamp >= expireTimestampMs.getAsLong()) {
+                                expiredPartitions.add(addOffsetCommitTombstone(groupId, topic, partition, records));
+                            }
+                            // Current version with no per partition retention
+                        } else if (currentTimestamp - expirationCondition.baseTimestamp.apply(offsetAndMetadata) >= offsetsRetentionMs) {
+                            expiredPartitions.add(addOffsetCommitTombstone(groupId, topic, partition, records));
+                        } else {
+                            hasAllOffsetsExpired.set(false);
+                        }
+                    });
+                }
+            });
+            log.debug("[GroupId {}] Expiring offsets: {}", groupId, expiredPartitions);
+
+            if (hasAllOffsetsExpired.get()) {
+                // All offsets were expired for this group. Remove the group.
+                return Optional.of(groupId);
+            }
+
+        } catch (GroupIdNotFoundException e) {
+            // groups in offsets should exist.
+            log.warn("GroupId {} should exist.", groupId);
+        }
```

Comment:
Given that the group should exist based on how we use this method, I think that we could remove the try..catch and just let the exception go to the caller. We don't need to catch it.